vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js

```diff
@@ -78,69 +132,176 @@
-                *  */
-                datepicker.$connector.regex = testString.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&').replace(datepicker.$connector.dayPart.initial, "(\\d{1,2})").replace(datepicker.$connector.monthPart.initial, "(\\d{1,2})").replace(datepicker.$connector.yearPart.initial, "(\\d{4})");
-
-                datepicker.i18n.formatDate = tryCatchWrapper(function (date) {
+                 * regex will be the date, so that:
+                 * - day-part is '(\d{1,2})' (1 or 2 digits),
+                 * - month-part is '(\d{1,2})' (1 or 2 digits),
+                 * - year-part is '(\d{1,4})' (1 to 4 digits)
+                 *
+                 * and everything else is left as is.
+                 * For example, us date "10/20/2010" => "(\d{1,2})/(\d{1,2})/(\d{1,4})".
+                 *
+                 * The sorting part solves that which part is which (for example,
+                 * here the first part is month, second day and third year)
+                 */
+                datepicker.$connector.regex = testString.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&')
+                    .replace(dayPart.initial, "(\\d{1,2})")
+                    .replace(monthPart.initial, "(\\d{1,2})")
+                    .replace(yearPart.initial, "(\\d{1,4})");
+
+                /**
+                 * @param {DateHash} date
+                 * @returns {number}
+                 */
+                let indexOfYearForFormattedDate = function (date) {
+                    let adjust = 0;
+                    for (let i = 1; i < 4; i++) {
+                        const part = parts[i - 1];
+                        // Adjust the index if the day or month (before the year)
+                        // have less digits than in the initial testDate so that
+                        // this works also with single digit day or month.
+                        // Day or month of less than 10 may still be 2 digit prefixed
+                        // with 0 depending on locale.
+                        if (part === dayPart && date.dayMatch) {
+                            adjust -= 2 - date.dayMatch.length;
+                        } else if (part === monthPart && date.monthMatch) {
+                            adjust -= 2 - date.monthMatch.length;
+                        } else if (part === yearPart) {
+                            return yearPart.index + adjust;
+                        }
+                    }
+                };
+                indexOfYearForFormattedDate = tryCatchWrapper(indexOfYearForFormattedDate);
+
+                /**
+                 * Prefixes the year in the given date string with zeros if
+                 * needed to make it four digits
+                 * @param {string} formattedDate Date formatted with Date.toLocaleDateString()
+                 * @returns {string}
+                 */
+                let yearToFourDigitInFormattedDate = function (formattedDate) {
```

Comment:
A new helper method to prefix the year in a formatted  date string with zeros to make it 4 digits (in case the year is less than 1000) since there is no other way to enforce a year to be formatted with 4 digits when formatting with `Date.toLocaleDateString()` (in all locales).